### PR TITLE
feat(router): add attach dispute evidence api

### PR DESCRIPTION
**Pull request by sai-harsha-vardhan**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Add support for attach evidence api which allows to upload a file, gets the file_id and attaches that particular file_id to dispute with evidence_type

### Additional Changes

- [x] This PR modifies the API contract
- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
Tested Manually
image

image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/disputes.rs [hunk 1]
```diff
@@ -304,3 +305,75 @@ pub async fn submit_evidence(
...
+        .parse_value("DisputeEvidence")
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Error while parsing dispute evidence record")?;
+    let updated_dispute_evidence = transformers::update_dispute_evidence(
+        dispute_evidence,
+        attach_evidence_request.evidence_type,
+        file_id,
+    );
+    let update_dispute = storage_models::dispute::DisputeUpdate::EvidenceUpdate {
+        evidence: serde_json::to_value(updated_dispute_evidence)
```
Comment (vspecky) [inline crates/router/src/core/disputes.rs:367]:
```suggestion
        evidence: utils::Encode::<DisputeEvidence>::encode_to_value(&updated_dispute_evidence)
```

File: crates/router/src/routes/disputes/utils.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        Some("refund_policy") => Some(disputes::EvidenceType::RefundPolicy),
+        Some("service_documentation") => Some(disputes::EvidenceType::ServiceDocumentation),
+        Some("shipping_documentation") => Some(disputes::EvidenceType::ShippingDocumentation),
+        Some("invoice_showing_distinct_transactions") => {
+            Some(disputes::EvidenceType::InvoiceShowingDistinctTransactions)
+        }
+        Some("recurring_transaction_agreement") => {
+            Some(disputes::EvidenceType::RecurringTransactionAgreement)
+        }
+        Some("uncategorized_file") => Some(disputes::EvidenceType::UncategorizedFile),
```
Comment (vspecky) [inline crates/router/src/routes/disputes/utils.rs:29]: Consider deserializing here

File: crates/router/src/routes/disputes/utils.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            _ => (),
+        }
+    }
+    let evidence_type = option_evidence_type.get_required_value("evidence_type")?;
+    let file = match file_content {
+        Some(valid_file_content) => valid_file_content.concat().to_vec(),
+        None => Err(errors::ApiErrorResponse::MissingFile)
+            .into_report()
+            .attach_printable("Missing / Invalid file in the request")?,
+    };
```
Comment (vspecky) [inline crates/router/src/routes/disputes/utils.rs:79]:
```suggestion
    let file = file_content
        .get_required_value("file_content")
        ...
        .concat()
        .to_vec()
```

File: crates/router/src/routes/disputes/utils.rs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    //Get and validate file size
+    let file_size: i32 = file
+        .len()
+        .try_into()
+        .into_report()
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("File size error")?;
```
Comment (vspecky) [inline crates/router/src/routes/disputes/utils.rs:86]: I just noticed this now but is there a good reason for the file size to be signed? If not, can we consider just making it a `usize`? (can take this up later)

File: crates/router/src/routes/disputes/utils.rs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    // Check if empty file and throw error
+    if file_size <= 0 {
+        Err(errors::ApiErrorResponse::MissingFile)
+            .into_report()
+            .attach_printable("Missing / Invalid file in the request")?
```
Comment (vspecky) [inline crates/router/src/routes/disputes/utils.rs:91]: use `utils::when()`

**Review by vspecky — CHANGES_REQUESTED** [review]

**Review by vspecky — DISMISSED** [review]

File: crates/router/src/core/metrics.rs [hunk 6]
```diff
@@ -26,3 +26,8 @@ counter_metric!(
     EVIDENCE_SUBMISSION_DISPUTE_STATUS_VALIDATION_FAILURE_METRIC,
     GLOBAL_METER
 ); //No. of status validation fialures while submitting evidence for a dispute
+   //No. of status validation fialures while attaching evidence for a dispute
```
Comment (SanchithHegde) [inline crates/router/src/core/metrics.rs:29]:
```suggestion
); //No. of status validation failures while submitting evidence for a dispute
   //No. of status validation failures while attaching evidence for a dispute
```

File: crates/router/src/routes/disputes/utils.rs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                file_name = content_disposition.get_filename().map(String::from);
+                //Collect the file content and throw error if something fails
+                let mut file_data = Vec::new();
+                let mut stream = field.into_stream();
+                while let Some(chunk) = stream.next().await {
+                    match chunk {
+                        Ok(bytes) => file_data.push(bytes),
+                        Err(err) => Err(errors::ApiErrorResponse::InternalServerError)
+                            .into_report()
+                            .attach_printable(format!("{}{}", "File parsing error: ", err))?,
```
Comment (SanchithHegde) [inline crates/router/src/routes/disputes/utils.rs:52]:
```suggestion
                            .attach_printable_lazy(|| format!("File parsing error: {err}"))?,
```

`attach_printable_lazy()` is preferred whenever there's a function call (in this case `format!()`) involved.

File: crates/router/src/services/api.rs [hunk 8]
```diff
@@ -450,7 +450,7 @@ async fn handle_response(
         .await
 }
 
-#[derive(Debug, Eq, PartialEq)]
+#[derive(Debug, Eq, PartialEq, Clone)]
```
Comment (SanchithHegde) [inline crates/router/src/services/api.rs:453]: What's the need of the response to be clonable?

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

File: crates/router/src/services/api.rs (hunk 8 above)
Comment (sai-harsha-vardhan) [reply crates/router/src/services/api.rs:453]: Am using it here - https://github.com/juspay/hyperswitch/pull/1070/files#diff-9cffb95d59254b4103a363d800294d393f2ab62e7933a68b65904c90132be933R350

Comment (SanchithHegde) [reply crates/router/src/services/api.rs:453]: Can't you clone only the `file_id` in that case? Something like this?

```rust
    let file_id = match create_file_response {
        services::ApplicationResponse::Json(ref res) => res.file_id.clone(),
        _ => Err(errors::ApiErrorResponse::InternalServerError)
            .into_report()
            .attach_printable("Unexpected response received from files create core")?,
    };
```

File: migrations/2023-05-05-112013_add_evidence_col_in_dispute/up.sql [hunk 9]
```diff
@@ -0,0 +1,3 @@
+-- Your SQL goes here
+ALTER TABLE dispute
+ADD COLUMN evidence JSONB NOT NULL DEFAULT '{}'::JSONB;
```
Comment (jarnura) [inline migrations/2023-05-05-112013_add_evidence_col_in_dispute/up.sql:3]: Remove not null constraints

**Review by jarnura — CHANGES_REQUESTED** [review]

File: migrations/2023-05-05-112013_add_evidence_col_in_dispute/up.sql (hunk 9 above)
Comment (sai-harsha-vardhan) [reply migrations/2023-05-05-112013_add_evidence_col_in_dispute/up.sql:3]: It's backward compatible as there is no dispute record yet!

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**